File: backend/app/schemas/ai.py
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field

from app.schemas.task import Priority


AiSuggestionType = Literal["generate_tasks", "split_task", "improve_task"]


class SuggestedTask(BaseModel):
    title: str = Field(min_length=1, max_length=120)
    description: str | None = Field(default=None, max_length=4000)
    priority: Priority | None = None
    tags: list[str] = Field(default_factory=list)


class GenerateTasksRequest(BaseModel):
    goal: str = Field(min_length=1, max_length=4000)


class GenerateTasksResult(BaseModel):
    tasks: list[SuggestedTask]


class SplitTaskRequest(BaseModel):
    text: str = Field(min_length=1, max_length=4000)


class SplitTaskResult(BaseModel):
    subtasks: list[SuggestedTask]


class ImproveTaskRequest(BaseModel):
    title: str = Field(min_length=1, max_length=120)
    description: str | None = Field(default=None, max_length=4000)


class ImproveTaskResult(BaseModel):
    title: str = Field(min_length=1, max_length=120)
    description: str | None = Field(default=None, max_length=4000)


class AiSuggestionResponse(BaseModel):
    id: int | None = None
    type: AiSuggestionType
    status: Literal["pending"] = "pending"
    provider: Literal["openrouter"] = "openrouter"
    model: str
    mock: bool
    result: GenerateTasksResult | SplitTaskResult | ImproveTaskResult


class AiSettingsRead(BaseModel):
    provider: Literal["openrouter"] = "openrouter"
    selected_model: str | None = None
    is_configured: bool
    mock: bool
    masked_api_key: str | None = None


class AiSettingsUpdate(BaseModel):
    api_key: str = Field(min_length=1, max_length=4000)
    selected_model: str = Field(min_length=1, max_length=200)


class AiSuggestionRead(BaseModel):
    id: int
    type: AiSuggestionType
    request_text: str
    response_payload: dict[str, Any]
    model: str
    status: Literal["pending", "applied", "cancelled"]
    created_at: datetime


class ApplySuggestionRequest(BaseModel):
    suggestion_id: int
    selected_indexes: list[int] | None = None

File: frontend/src/types/ai.ts
import type { TaskPriority } from './task'

export type AiSuggestedTask = {
  title: string
  description: string | null
  priority: TaskPriority | null
  tags: string[]
}

export type GenerateTasksResponse = {
  id: number | null
  type: 'generate_tasks'
  status: 'pending'
  provider: 'openrouter'
  model: string
  mock: boolean
  result: {
    tasks: AiSuggestedTask[]
  }
}

export type SplitTaskResponse = {
  id: number | null
  type: 'split_task'
  status: 'pending'
  provider: 'openrouter'
  model: string
  mock: boolean
  result: {
    subtasks: AiSuggestedTask[]
  }
}

export type ImproveTaskResponse = {
  id: number | null
  type: 'improve_task'
  status: 'pending'
  provider: 'openrouter'
  model: string
  mock: boolean
  result: {
    title: string
    description: string | null
  }
}

export type AiSettings = {
  provider: 'openrouter'
  selected_model: string | null
  is_configured: boolean
  mock: boolean
  masked_api_key: string | null
}

export type AiSuggestionHistoryItem = {
  id: number
  type: 'generate_tasks' | 'split_task' | 'improve_task'
  request_text: string
  response_payload: Record<string, unknown>
  model: string
  status: 'pending' | 'applied' | 'cancelled'
  created_at: string
}
